File: app/src/main/java/com/elewise/ldmobile/rest/RestHelper.java
package com.elewise.ldmobile.rest;

import android.util.Log;

import com.elewise.ldmobile.BuildConfig;
import com.elewise.ldmobile.api.*;
import com.elewise.ldmobile.model.ProcessType;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.Body;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

public class RestHelper {

    private static RestHelper restHelper = null;
    private static RestApi service = null;

    public static RestHelper getInstance(String baseUrl) {
        if (restHelper == null) {
            restHelper = new RestHelper(baseUrl);
        }
        return restHelper;
    }

    private OkHttpClient createOkHttpClient() {
        if (BuildConfig.DEBUG) {
            return new OkHttpClient().newBuilder()
                    .connectTimeout(20, TimeUnit.SECONDS)
                    .writeTimeout(20, TimeUnit.SECONDS)
                    .readTimeout(60, TimeUnit.SECONDS)
//                    .addInterceptor(authInterceptor())
                    .addInterceptor(createHttpLoggingInterceptor())
                    .build();
        } else {
            return new OkHttpClient().newBuilder()
                    .connectTimeout(20, TimeUnit.SECONDS)
                    .writeTimeout(20, TimeUnit.SECONDS)
                    .readTimeout(60, TimeUnit.SECONDS)
                    .addInterceptor(createHttpLoggingInterceptor())
//                    .addInterceptor(authInterceptor())
                    .build();
        }
    }

    private HttpLoggingInterceptor createHttpLoggingInterceptor() {
        HttpLoggingInterceptor httpLoggingInterceptor = new HttpLoggingInterceptor(message -> {
            Log.e("httpLog", message);
        });
        httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        return httpLoggingInterceptor;
    }

    public RestHelper(String baseUrl) {
        OkHttpClient client = createOkHttpClient();


        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(baseUrl)
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build();
        service = retrofit.create(RestApi.class);
    }

    public ParamAuthorizationResponse getAuthorizationTokenSync(String userName, String password) throws IOException {
        Call<ParamAuthorizationResponse> callRes = service.getAuthorizationToken(new ParamAuthorizationRequest(userName, password));
        return callRes.execute().body();
    }


    public ParamDocumentsResponse getDocumentsSync(String token, int size, int from, ProcessType processType, FilterData[] filterData) throws IOException {
        Call<ParamDocumentsResponse> callRes = service.getDocuments(new ParamDocumentsRequest(token, size, from, processType.getType(), filterData));
        return callRes.execute().body();
    }

    public ParamDocumentDetailsResponse getDocumentDetailsSync(String token, String doc_type, Integer doc_id) throws IOException {
        Call<ParamDocumentDetailsResponse> callRes = service.getDocumentDetails(new ParamDocumentDetailsRequest(token, doc_id, doc_type));
        return callRes.execute().body();
    }

    public ParamFilterSettingsResponse getFilterSettings(String token) throws IOException {
        Call<ParamFilterSettingsResponse> callRes = service.getFilterSettings(new ParamFilterSettingsRequest(token));
        return callRes.execute().body();
    }

    public byte[] getFile(ParamGetFileRequest request) throws IOException {
        Call<byte[]> callRes = service.getFile(request);
        return callRes.execute().body();
    }

    public ParamSaveFileSignResponse saveFileSign(ParamSaveFileSignRequest request) throws IOException {
        Call<ParamSaveFileSignResponse> callRes = service.saveFileSign(request);
        return callRes.execute().body();
    }

    public ParamExecDocumentResponse execDocument(ParamExecDocumentRequest request) throws IOException {
        Call<ParamExecDocumentResponse> callRes = service.execDocument(request);
        return callRes.execute().body();
    }
}

File: app/src/main/java/com/elewise/ldmobile/rest/RestHelper.kt
package com.elewise.ldmobile.rest

import android.util.Log
import com.elewise.ldmobile.BuildConfig
import com.elewise.ldmobile.api.*
import com.elewise.ldmobile.api.ParamExecOperationResponse
import com.elewise.ldmobile.api.ParamGetFileResponse
import com.elewise.ldmobile.model.ProcessType
import com.jakewharton.retrofit2.adapter.kotlin.coroutines.CoroutineCallAdapterFactory
import kotlinx.coroutines.Deferred
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.IOException
import java.util.concurrent.TimeUnit


class RestHelper(baseUrl: String) {
    private val restApi: RestApi

    init {
        val client = createOkHttpClient()
        val retrofit = Retrofit.Builder()
                .baseUrl(baseUrl)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(CoroutineCallAdapterFactory())
                .client(client)
                .build()
        restApi = retrofit.create(RestApi::class.java)
    }

    private fun createOkHttpClient(): OkHttpClient {
        return if (BuildConfig.DEBUG) {
            OkHttpClient().newBuilder()
                    .connectTimeout(20, TimeUnit.SECONDS)
                    .writeTimeout(20, TimeUnit.SECONDS)
                    .readTimeout(60, TimeUnit.SECONDS)
                    .addInterceptor(createHttpLoggingInterceptor())
                    .build()
        } else {
            OkHttpClient().newBuilder()
                    .connectTimeout(20, TimeUnit.SECONDS)
                    .writeTimeout(20, TimeUnit.SECONDS)
                    .readTimeout(60, TimeUnit.SECONDS)
                    .addInterceptor(createHttpLoggingInterceptor())
                    .build()
        }
    }

    private fun createHttpLoggingInterceptor(): HttpLoggingInterceptor {
        val httpLoggingInterceptor = HttpLoggingInterceptor { message ->
            Log.e("httpLog", message)
        }
        httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY)
        return httpLoggingInterceptor
    }

    @Throws(IOException::class)
    fun getAuthorizationTokenSync(userName: String, password: String, deviceId: String): Deferred<Response<ParamAuthorizationResponse?>> {
        return restApi.getAuthorizationToken(ParamAuthorizationRequest(userName, password, deviceId))
    }

    @Throws(IOException::class)
    fun tokenActivityCheck(token: String): Deferred<Response<ParamTokenActivityCheckResponse?>> {
        return restApi.tokenActivityCheck(ParamTokenActivityCheckRequest(token))
    }

    @Throws(IOException::class)
    fun getDocumentsSync(token: String, size: Int, from: Int, processType: ProcessType, filterData: List<FilterData>): Deferred<Response<ParamDocumentsResponse>> {
        return restApi.getDocuments(ParamDocumentsRequest(token, size, from, processType.type, filterData))
    }

    @Throws(IOException::class)
    fun getDocumentDetailsSync(token: String, doc_id: Int): Deferred<Response<ParamDocumentDetailsResponse>> {
        return restApi.getDocumentDetails(ParamDocumentDetailsRequest(token, doc_id))
    }

    @Throws(IOException::class)
    fun getFilterSettings(token: String): Deferred<Response<ParamFilterSettingsResponse>> {
        return restApi.getFilterSettings(ParamFilterSettingsRequest(token))
    }

    @Throws(IOException::class)
    fun getFile(request: ParamGetFileRequest?): Deferred<Response<ParamGetFileResponse?>> {
        return restApi.getFile(request)
    }

    @Throws(IOException::class)
    fun saveFileSign(request: ParamSaveFileSignRequest?): Deferred<Response<ParamSaveFileSignResponse?>> {
        return restApi.saveFileSign(request)
    }

    @Throws(IOException::class)
    fun execDocument(request: ParamExecDocumentRequest?): Deferred<Response<ParamExecDocumentResponse?>> {
        return restApi.execDocument(request)
    }

    @Throws(IOException::class)
    fun execOperation(request: ParamExecOperationRequest): Deferred<Response<ParamExecOperationResponse?>> {
        return restApi.execOperation(request)
    }

    companion object {
        fun createNewInstance(baseUrl: String): RestHelper {
            return RestHelper(baseUrl)
        }
    }
}